File: easy/0566_reshape_the_matrix/solution.cpp
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>>
  matrixReshape(std::vector<std::vector<int>> &mat, int r, int c) {
    const size_t cols = mat[0].size();

    if (static_cast<size_t>(r * c) != mat.size() * cols)
      return mat;

    std::vector<std::vector<int>> reshapedMat(r, std::vector<int>(c, 0));

    for (size_t i = 0; i < static_cast<size_t>(r * c); ++i)
      reshapedMat[i / c][i % c] = mat[i / cols][i % cols];

    return reshapedMat;
  }
};

File: easy/0566_reshape_the_matrix/Solution.java
class Solution {
  public int[][] matrixReshape(int[][] mat, int r, int c) {
    final int cols = mat[0].length;

    if (r * c != mat.length * cols) {
      return mat;
    }

    final int[][] reshapedMat = new int[r][c];

    for (int i = 0; i < r * c; ++i) {
      reshapedMat[i / c][i % c] = mat[i / cols][i % cols];
    }

    return reshapedMat;
  }
}
